bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java

```diff
@@ -471,13 +463,60 @@ public void start(BundleContext context) throws Exception {
 		Hashtable<String, String> properties = new Hashtable<>(2);
 		properties.put(DebugOptions.LISTENER_SYMBOLICNAME, PI_RESOURCES);
 		debugRegistration = context.registerService(DebugOptionsListener.class, Policy.RESOURCES_DEBUG_OPTIONS_LISTENER, properties);
-		// Remember workspace before opening, to
-		// make it easier to debug cases where open() is failing.
-		workspace = new Workspace();
-		IStatus result = workspace.open(null);
-		if (!result.isOK())
-			getLog().log(result);
-		workspaceRegistration = context.registerService(IWorkspace.class, workspace, null);
+		instanceLocationTracker = new ServiceTracker<>(context,
```

Comment:
Could we extract this inner class to something like below?
That would simplify understanding of the code I think.

```
diff --git a/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java b/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java
index 6b1c549..49e0a2b 100644
--- a/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java
+++ b/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java
@@ -47,2 +47,3 @@
 public final class ResourcesPlugin extends Plugin {
+
 	/**
@@ -468,52 +469,3 @@
 						Location.SERVICE_PROPERTY_URL)),
-				new ServiceTrackerCustomizer<>() {
-
-					private ServiceRegistration<IWorkspace> workspaceRegistration;
-
-					@Override
-					public Workspace addingService(ServiceReference<Location> reference) {
-						if (workspace == null) {
-							Location location = context.getService(reference);
-							if (location != null) {
-								// Remember workspace before opening, to
-								// make it easier to debug cases where open() is failing.
-								workspace = new Workspace();
-								IStatus result;
-								try {
-									result = workspace.open(null);
-									if (!result.isOK())
-										getLog().log(result);
-									workspaceRegistration = context.registerService(IWorkspace.class, workspace, null);
-									return workspace;
-								} catch (CoreException e) {
-									getLog().log(e.getStatus());
-								}
-							}
-						}
-						return null;
-					}
-
-					@Override
-					public void modifiedService(ServiceReference<Location> reference, Workspace service) {
-						// nothing to care about here as we already tracking the service with the
-						// properties of interest
-					}
-
-					@Override
-					public void removedService(ServiceReference<Location> reference, Workspace service) {
-						if (service == workspace) {
-							workspaceRegistration.unregister();
-							try {
-								service.close(null);
-								// Forget workspace only if successfully closed, to
-								// make it easier to debug cases where close() is failing.
-								workspace = null;
-							} catch (CoreException e) {
-								getLog().log(e.getStatus());
-							} finally {
-								context.ungetService(reference);
-							}
-						}
-					}
-				});
+				new WorkspaceInit(context));
 		instanceLocationTracker.open();
@@ -521,2 +473,63 @@
 
+	/**
+	 * Creates and opens singleton workspace instance for this session
+	 */
+	private final class WorkspaceInit implements ServiceTrackerCustomizer<Location, Workspace> {
+		private final BundleContext context;
+		private ServiceRegistration<IWorkspace> workspaceRegistration;
+
+		private WorkspaceInit(BundleContext context) {
+			this.context = context;
+		}
+
+		@Override
+		public Workspace addingService(ServiceReference<Location> reference) {
+			if (workspace != null) {
+				return null;
+			}
+			Location location = context.getService(reference);
+			if (location == null) {
+				return null;
+			}
+			// Remember workspace before opening, to
+			// make it easier to debug cases where open() is failing.
+			workspace = new Workspace();
+			try {
+				IStatus result = workspace.open(null);
+				if (!result.isOK()) {
+					getLog().log(result);
+				}
+				workspaceRegistration = context.registerService(IWorkspace.class, workspace, null);
+				return workspace;
+			} catch (CoreException e) {
+				getLog().log(e.getStatus());
+			}
+			return null;
+		}
+
+		@Override
+		public void modifiedService(ServiceReference<Location> reference, Workspace service) {
+			// nothing to care about here as we already tracking the service with the
+			// properties of interest
+		}
+
+		@Override
+		public void removedService(ServiceReference<Location> reference, Workspace service) {
+			if (service != workspace) {
+				return;
+			}
+			workspaceRegistration.unregister();
+			try {
+				service.close(null);
+				// Forget workspace only if successfully closed, to
+				// make it easier to debug cases where close() is failing.
+				workspace = null;
+			} catch (CoreException e) {
+				getLog().log(e.getStatus());
+			} finally {
+				context.ungetService(reference);
+			}
+		}
+	}
+
 }
```

Reply:
Done.